dpctl/memory/_sycl_usm_array_interface_utils.pxi

```diff
@@ -0,0 +1,187 @@
+    if (nd > 0):
+        if (ary_strides is None):
+            nelems = 1
+            for si in ary_shape:
+                sh_i = int(si)
+                if (sh_i <= 0):
+                    raise ValueError("Array shape elements need to be positive")
+                nelems = nelems * sh_i
+            return (ary_offset, nelems * itemsize)
+        else:
+            min_disp = ary_offset
+            max_disp = ary_offset
+            for i in range(nd):
+                str_i = int(ary_strides[i])
+                sh_i = int(ary_shape[i])
+                if (str_i > 0):
+                    max_disp += str_i * (sh_i - 1)
+                else:
+                    min_disp += str_i * (sh_i - 1);
+            return (min_disp, (max_disp - min_disp + 1) * itemsize)
+    elif (nd == 0):
+        return (ary_offset, itemsize)
+    else:
+        raise ValueError("Array dimensions can not be negative")
+
+
+cdef class _USMBufferData:
+    """
+    Internal data struct populated from parsing
+    `__sycl_usm_array_interface__` dictionary
+    """
+    cdef DPCTLSyclUSMRef p
+    cdef int writeable
+    cdef object dt
+    cdef Py_ssize_t itemsize
+    cdef Py_ssize_t nbytes
+    cdef SyclQueue queue
+
+    @staticmethod
+    cdef _USMBufferData from_sycl_usm_ary_iface(dict ary_iface):
+        cdef object ary_data_tuple = ary_iface.get('data', None)
+        cdef object ary_typestr = ary_iface.get('typestr', None)
+        cdef object ary_shape = ary_iface.get('shape', None)
+        cdef object ary_strides = ary_iface.get('strides', None)
+        cdef object ary_syclobj = ary_iface.get('syclobj', None)
+        cdef Py_ssize_t ary_offset = ary_iface.get('offset', 0)
+        cdef int ary_version = ary_iface.get('version', 0)
+        cdef Py_ssize_t arr_data_ptr = 0
+        cdef DPCTLSyclUSMRef memRef = NULL
+        cdef Py_ssize_t itemsize = -1
+        cdef int writeable = -1
+        cdef int nd = -1
+        cdef DPCTLSyclQueueRef QRef = NULL
+        cdef object dt
+        cdef _USMBufferData buf
+        cdef SyclDevice dev
+        cdef SyclContext ctx
+
+        if ary_version != 1:
+            raise ValueError(("__sycl_usm_array_interface__ is malformed:"
+                              " dict('version': {}) is unexpected."
+                              " The only recognized version is 1.").format(ary_version))
+        if not ary_data_tuple or len(ary_data_tuple) != 2:
+            raise ValueError("__sycl_usm_array_interface__ is malformed:"
+                             " 'data' field is required, and must be a tuple"
+                             " (usm_pointer, is_writeable_boolean).")
+        arr_data_ptr = <Py_ssize_t>ary_data_tuple[0]
+        writeable = 1 if ary_data_tuple[1] else 0
+        # Check that memory and syclobj are consistent:
+        # (USM pointer is bound to this sycl context)
+        memRef = <DPCTLSyclUSMRef>arr_data_ptr
+        QRef = get_queue_ref_from_ptr_and_syclobj(memRef, ary_syclobj)
```

Comment:
`QRef` leaks when you raise an exception between here and its use on L 185.

Reply:
Fixed that. Added calls to `DPCTLQueue_Delete` before raising the exception. Thanks for catching it.